### PR TITLE
Nvm mmap

### PR DESCRIPTION
**Pull request by BenStoler**
Initial Implementation of open/mmap for a persistent memory file. Use via:

`klee --posix-runtime --libc=uclibc BITCODE --sym-pmem FILENAME SIZE_IN_BYTES`

Example:
```
int main() {
  int fd = open("data", O_RDWR | O_CREAT | O_TRUNC, 0600);
  int size = 128;
  char *addr = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
  if (addr == (void*) -1 ){
          exit(1);
  }

  int a, b;
  klee_make_symbolic(&a, sizeof(a), "a");
  klee_assume(0 <= a & a < size);
  klee_make_symbolic(&b, sizeof(b), "b");
  klee_assume(0 <= b & b < size);

  //will only succeed if a and b on same cache-line
  addr[a] = 'a';
  _mm_clwb(&addr[b]);
  _mm_sfence();

  klee_pmem_check_persisted(&addr[a], sizeof(addr[a]));
  
  munmap(addr, size);
  printf("Program done\n");
  return 0;
  
}

```

This can then be called like:
`klee --posix-runtime --libc=uclibc FILE.bc --sym-pmem data 4096`

Remaining limitations:
- For some reason, having the stat64 object in the disk file being symbolic results in two extra paths being generated, but they can just be ignored for now
- For now, only works with a single file being passed in, rather than supporting multiple persistent file devices; can be changed easily in the future
- The name of the persistent Memory Object (when using klee_pmem_mark_persistent) for now is hard coded to be "data". It's a FIXME to change that name to be the corresponding filename, at some point in the future.
- Also in the future, trying to handle flags passed in to mmap better could be nice.
- For now, munmap doesn't do anything, since mmap doesn't perform any new allocation, instead just returning a pointer into the data.
- More testing of edge conditions with mmap w.r.t. length and offset permutations could be beneficial

### CODE REVIEW

File: runtime/POSIX/fd.h [hunk 1]
```diff
@@ -93,7 +98,8 @@ extern exe_sym_env_t __exe_env;
 
 void klee_init_fds(unsigned n_files, unsigned file_length,
                    unsigned stdin_length, int sym_stdout_flag,
-                   int do_all_writes_flag, unsigned max_failures);
+                   int do_all_writes_flag, unsigned max_failures,
+				   char* sym_pmem_filename, unsigned sym_pmem_size);
```
Comment (iangneal) [inline runtime/POSIX/fd.h]: At some point, we will want to support multiple pmem files, but for now this should be fine.

File: runtime/POSIX/fd_init.c [hunk 2]
```diff
@@ -109,7 +113,8 @@ static unsigned __sym_uint32(const char *name) {
    max_failures: maximum number of system call failures */
 void klee_init_fds(unsigned n_files, unsigned file_length,
                    unsigned stdin_length, int sym_stdout_flag,
-                   int save_all_writes_flag, unsigned max_failures) {
+                   int save_all_writes_flag, unsigned max_failures,
+				   char* sym_pmem_filename, unsigned sym_pmem_size) {
```
Comment (iangneal) [inline runtime/POSIX/fd_init.c]: Your spacing is a little funny, are you using tabs?

File: runtime/POSIX/stubs.c [hunk 3]
```diff
@@ -590,7 +596,30 @@ ssize_t readahead(int fd, off64_t *offset, size_t count) {
 
 void *mmap(void *start, size_t length, int prot, int flags, int fd, off_t offset) __attribute__((weak));
 void *mmap(void *start, size_t length, int prot, int flags, int fd, off_t offset) {
```
Comment (iangneal) [inline runtime/POSIX/stubs.c:666]: PMDK actually does use the `void *start` argument, so you should support it.

File: runtime/POSIX/stubs.c [hunk 4]
```diff
@@ -590,7 +596,30 @@ ssize_t readahead(int fd, off64_t *offset, size_t count) {
...
+			  printf("mmapping the pmem file at offset %lld, with length %lld (rounded from %lld).\n", 
+					  offset, actual_length, length);
+			  printf("mmap at address %p\n", (void*) (df->contents + offset));
+#endif
+			  return (void*)  (df->contents + offset);
+		  }
+	  }
+  } else {
+	  klee_warning("ignoring (EPERM), since non file-backed or not a pmem file");
+  }
```
Comment (iangneal) [inline runtime/POSIX/stubs.c]: Nitpick, but let's be consistent with KLEE's existing spacing style if possible.

**Review by iangneal — CHANGES_REQUESTED** [review]
Made a couple of conflicts. Also, please resolve merge conflicts.

**Comment by iangneal** [comment]
Thank you for the in-depth description as well. That was very helpful.

File: runtime/POSIX/stubs.c (hunk 4 above)
Comment (BGR360) [reply runtime/POSIX/stubs.c]: agreed

**Review by iangneal — APPROVED** [review]
